File: MultiBoxJavaCore/src/sk/hackcraft/multibox/model/ServerLibraryShadow.java
package sk.hackcraft.multibox.model;

import java.util.LinkedList;
import java.util.List;

import sk.hackcraft.multibox.net.ServerInterface;
import sk.hackcraft.multibox.net.ServerInterface.ServerInterfaceEventAdapter;
import sk.hackcraft.util.MessageQueue;

public class ServerLibraryShadow implements Library
{	
	private final ServerInterface serverInterface;
	private final MessageQueue messageQueue;
	
	private ServerListener serverListener;
	
	private final List<LibraryEventListener> libraryListeners;

	public ServerLibraryShadow(ServerInterface serverInterface, MessageQueue messageQueue)
	{
		this.serverInterface = serverInterface;
		this.messageQueue = messageQueue;
		
		this.serverListener = new ServerListener();
		
		libraryListeners = new LinkedList<Library.LibraryEventListener>();
	}
	
	@Override
	public void init()
	{
		serverInterface.registerEventListener(serverListener);
	}

	@Override
	public void close()
	{
		serverInterface.unregisterEventListener(serverListener);
	}

	@Override
	public void requestItem(long id)
	{
		serverInterface.requestLibraryItem(id);
	}

	@Override
	public void registerLibraryEventListener(LibraryEventListener listener)
	{
		libraryListeners.add(listener);
	}

	@Override
	public void unregisterLibraryEventListener(LibraryEventListener listener)
	{
		libraryListeners.remove(listener);
	}
	
	private void notifyLibraryItemReceived(final LibraryItem item)
	{
		for (final Library.LibraryEventListener listener : libraryListeners)
		{
			messageQueue.post(new Runnable()
			{
				@Override
				public void run()
				{
					listener.onItemReceived(item);
				}
			});
		}
	}
	
	private class ServerListener extends ServerInterfaceEventAdapter
	{
		@Override
		public void onLibraryItemReceived(LibraryItem item)
		{
			notifyLibraryItemReceived(item);
		}
	}
}

File: MultiboxClientCSharpCore/Model/ServerLibraryShadow.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using CSharpNetInterface.Util;
using MultiboxClientCSharpCore.Net;
using MultiboxClientCSharpCore.Model.LibraryEventArgs;

namespace MultiboxClientCSharpCore.Model
{
	class ServerLibraryShadow : ILibrary
	{
		private readonly IServerInterface serverInterface;
		private readonly IMessageQueue messageQueue;

		public event EventHandler<LibraryItemReceivedEventArgs> LibraryItemReceived;

		private readonly PeriodicWorkDispatcher stateChecker;

		private long lastRequestedId = LibraryConstants.RootDirectory;

		public ServerLibraryShadow(IServerInterface serverInterface, IMessageQueue messageQueue)
		{
			this.serverInterface = serverInterface;
			this.messageQueue = messageQueue;

			this.stateChecker = new PeriodicWorkDispatcher(messageQueue, 5000, UpdateLibrary);
		}

		private void OnLibraryItemReceived(object sender, MultiboxClientCSharpCore.Net.ServerInterfaceEventArgs.LibraryItemReceivedEventArgs eventArgs)
		{
			ILibraryItem item = eventArgs.LibraryItem;
			LibraryItemReceivedEventArgs libraryEventArgs = new LibraryItemReceivedEventArgs(item);

			messageQueue.Post(() => LibraryItemReceived(this, libraryEventArgs));
		}
				
		public void Init()
		{
			serverInterface.LibraryItemReceived += OnLibraryItemReceived;

			stateChecker.Start();
		}

		public void Close()
		{
			serverInterface.LibraryItemReceived -= OnLibraryItemReceived;

			stateChecker.Stop();
		}

		private void UpdateLibrary()
		{
			RequestItem(lastRequestedId);
		}

		public void RequestItem(long id)
		{
			serverInterface.RequestLibraryItem(id);

			this.lastRequestedId = id;
		}
	}
}
